File: scripts/extract-pdf-text.ts
#!/usr/bin/env tsx
/**
 * Extract text from downloaded PDFs for validation
 */
import * as fs from 'fs/promises'
import * as path from 'path'
import FormData from 'form-data'
import fetch from 'node-fetch'

async function extractPdfText() {
  const downloadDir = path.join(process.cwd(), 'downloaded-docs')
  const outputDir = path.join(process.cwd(), 'extracted-text')
  await fs.mkdir(outputDir, { recursive: true })

  const files = await fs.readdir(downloadDir)
  const pdfFiles = files.filter(f => f.endsWith('.pdf'))

  console.log(`Found ${pdfFiles.length} PDFs to extract`)

  const docWorkerUrl = process.env.DOC_WORKER_URL || 'http://localhost:8000'

  for (const pdfFile of pdfFiles) {
    try {
      console.log(`\n📄 Extracting: ${pdfFile}`)
      
      const pdfPath = path.join(downloadDir, pdfFile)
      const fileBuffer = await fs.readFile(pdfPath)
      
      // Call doc-worker V2
      const formData = new FormData()
      formData.append('file', fileBuffer, {
        filename: pdfFile,
        contentType: 'application/pdf',
      })

      const response = await fetch(`${docWorkerUrl}/extract/v2`, {
        method: 'POST',
        body: formData,
        headers: formData.getHeaders(),
      })

      if (!response.ok) {
        throw new Error(`Doc-worker returned ${response.status}: ${await response.text()}`)
      }

      const result = await response.json()
      
      // Save full extraction JSON
      const jsonPath = path.join(outputDir, pdfFile.replace('.pdf', '.json'))
      await fs.writeFile(jsonPath, JSON.stringify(result, null, 2))
      
      // Save plain text
      const textPath = path.join(outputDir, pdfFile.replace('.pdf', '.txt'))
      const fullText = result.items
        .map((item: any) => {
          const header = `\n=== Page ${item.page} | ${item.element_type} | ${item.section_path || 'no-section'} ===\n`
          let content = item.text
          if (item.has_verbatim && item.verbatim_block) {
            content += `\n\n[VERBATIM BLOCK]:\n${item.verbatim_block}\n`
          }
          return header + content
        })
        .join('\n\n')
      
      await fs.writeFile(textPath, fullText)
      
      console.log(`✅ Extracted ${result.items.length} blocks → ${textPath}`)
      console.log(`   📊 Verbatim blocks: ${result.items.filter((i: any) => i.has_verbatim).length}`)
    } catch (err) {
      console.error(`❌ Failed to extract ${pdfFile}:`, err)
    }
  }

  console.log(`\n✅ All extractions complete in: ${outputDir}`)
}

extractPdfText().catch(console.error)

File: scripts/extract-pdf-text.py
#!/usr/bin/env python3
"""
Simple PDF text extractor using PyPDF2 (no external dependencies)
"""
import sys
import json
from pathlib import Path

def extract_text_simple(pdf_path):
    """Extract text using basic Python libraries"""
    try:
        # Try PyPDF2 first
        import PyPDF2
        with open(pdf_path, 'rb') as file:
            reader = PyPDF2.PdfReader(file)
            text = ""
            for page in reader.pages:
                text += page.extract_text() + "\n"
            return text.strip()
    except ImportError:
        try:
            # Fallback to pdfplumber
            import pdfplumber
            with pdfplumber.open(pdf_path) as pdf:
                text = ""
                for page in pdf.pages:
                    text += page.extract_text() + "\n"
                return text.strip()
        except ImportError:
            print("❌ No PDF libraries available. Install: pip install PyPDF2")
            return None

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python extract-pdf-text.py <pdf_file>")
        sys.exit(1)
    
    pdf_path = sys.argv[1]
    if not Path(pdf_path).exists():
        print(f"❌ File not found: {pdf_path}")
        sys.exit(1)
    
    print(f"📄 Extracting text from: {pdf_path}")
    text = extract_text_simple(pdf_path)
    
    if text:
        print(f"✅ Extracted {len(text)} characters")
        print("📝 First 500 characters:")
        print("-" * 50)
        print(text[:500])
        print("-" * 50)
        
        # Save to file
        output_path = pdf_path.replace('.pdf', '_extracted.txt')
        with open(output_path, 'w', encoding='utf-8') as f:
            f.write(text)
        print(f"💾 Saved to: {output_path}")
    else:
        print("❌ Failed to extract text")
        sys.exit(1)



